File: PAT_Basic_Level_by_cpp/1125 子串与子列/1125 子串与子列.cpp
// 1125 子串与子列
//
// 实现原理：
// 子串是字符串中连续的一部分，子列是保持字符顺序的子集（可连续可不连续）。
// 给定字符串 S 和子列 P，找到 S 中包含 P 的最短子串；若解不唯一，输出起点最靠左的解。
//
// 关键步骤：
// 1. 枚举每个可能的起点 i（要求 S[i] == P[0]）。
// 2. 从 i 开始贪心匹配 P：按顺序在 S 中寻找 P 的每个字符，得到最早结束位置 j。
// 3. 窗口 [i, j-1] 即为以 i 为起点包含 P 的最短子串。
// 4. 记录长度最短的窗口；长度相同时保留起点最靠左的（严格小于才更新）。
//
// 复杂度分析：
// 时间复杂度：O(|S| * |P|)，最坏情况。
// 空间复杂度：O(1)，仅使用常数个变量。
#include <iostream>
#include <string>

using namespace std;

int main() {
    string s, p;
    if (!(cin >> s >> p)) return 0;

    int n = (int)s.size(), m = (int)p.size();
    int bestLen = n + 1, bestStart = 0;
    for (int i = 0; i < n; ++i) {
        if (s[i] != p[0]) continue;   // 起点必须匹配 P 的第一个字符
        int j = i, k = 0;
        while (j < n && k < m) {      // 贪心匹配 P
            if (s[j] == p[k]) ++k;
            ++j;
        }
        if (k == m) {                 // 匹配成功，窗口为 [i, j-1]
            int len = j - i;
            if (len < bestLen) { bestLen = len; bestStart = i; }  // 严格小于，保留最左解
        }
    }
    cout << s.substr(bestStart, bestLen) << endl;
    return 0;
}

File: PAT_Basic_Level_by_perl/1125 子串与子列/1125 子串与子列.pl
use strict;
use warnings;

# =====================================================================
# 题目：1125 子串与子列
# 难度：中等
# 知识点：双指针, 子序列匹配
# =====================================================================
#
# 【题目描述】
#   子串是字符串中连续的一部分，子列是保持顺序但不一定连续的子集。
#   给定字符串 S 和作为其子列的 P，找出 S 中包含 P 的最短子串；
#   若解不唯一，输出起点最靠左的解。
#
# 【实现原理】
#   1. 从起点 i 向右匹配 P，得到最右匹配位置 j。
#   2. 再从右往左收缩，找到包含 P 的最小结束位置 end。
#   3. 记录该子串长度，更新最短且起点靠左的解，然后从 end+1 继续。
#
# 【输入样例】
#   atpaaabpabttpcat
#   pat
# 【输出样例】
#   pabt
# =====================================================================

my $s = <STDIN>;
chomp $s;
my $p = <STDIN>;
chomp $p;

my $ls = length($s);
my $lp = length($p);

my $bestlen  = $ls + 1;
my $beststart = 0;
my $bestend   = 0;

my $i = 0;
while ($i < $ls) {
    # 从 i 向右找 P 的最右匹配位置 j
    my $j = $i;
    my $k = 0;
    while ($j < $ls && $k < $lp) {
        if (substr($s, $j, 1) eq substr($p, $k, 1)) {
            $k++;
        }
        $j++;
    }
    if ($k < $lp) {
        last;    # 无法再匹配 P，结束
    }

    # 从右向左收缩，找到包含 P 的最小结束位置 end
    my $end = $j - 1;
    $k = $lp - 1;
    while ($end >= $i) {
        if (substr($s, $end, 1) eq substr($p, $k, 1)) {
            $k--;
        }
        last if $k < 0;
        $end--;
    }

    # 记录更短的子串（起点从左向右枚举，自然靠左优先）
    if ($j - $end < $bestlen) {
        $bestlen  = $j - $end;
        $beststart = $end;
        $bestend   = $j;
    }

    $i = $end + 1;
}

print substr($s, $beststart, $bestend - $beststart), "\n";
